### Importación de datos



In [2]:
import pandas as pd

url = "https://raw.githubusercontent.com/alura-es-cursos/challenge1-data-science-latam/refs/heads/main/base-de-datos-challenge1-latam/tienda_1%20.csv"
url2 = "https://raw.githubusercontent.com/alura-es-cursos/challenge1-data-science-latam/refs/heads/main/base-de-datos-challenge1-latam/tienda_2.csv"
url3 = "https://raw.githubusercontent.com/alura-es-cursos/challenge1-data-science-latam/refs/heads/main/base-de-datos-challenge1-latam/tienda_3.csv"
url4 = "https://raw.githubusercontent.com/alura-es-cursos/challenge1-data-science-latam/refs/heads/main/base-de-datos-challenge1-latam/tienda_4.csv"

tienda = pd.read_csv(url)
tienda2 = pd.read_csv(url2)
tienda3 = pd.read_csv(url3)
tienda4 = pd.read_csv(url4)

tienda.head()

Producto Categoría del Producto    Precio  Costo de envío  \
0  Asistente virtual           Electrónicos  164300.0          6900.0   
1    Mesa de comedor                Muebles  192300.0          8400.0   
2      Juego de mesa               Juguetes  209600.0         15900.0   
3         Microondas      Electrodomésticos  757500.0         41000.0   
4   Silla de oficina                Muebles  335200.0         20200.0   

  Fecha de Compra         Vendedor Lugar de Compra  Calificación  \
0      16/01/2021      Pedro Gomez          Bogotá             4   
1      18/05/2022  Beatriz Morales        Medellín             1   
2      15/03/2021   Juan Fernandez       Cartagena             1   
3      03/05/2022   Juan Fernandez            Cali             4   
4      07/11/2020    Maria Alfonso        Medellín             5   

       Método de pago  Cantidad de cuotas       lat       lon  
0  Tarjeta de crédito                   8   4.60971 -74.08175  
1  Tarjeta de crédito                   4   6.25184 -75.56359  
2  Tarjeta de crédito                   1  10.39972 -75.51444  
3               Nequi                   1   3.43722 -76.52250  
4               Nequi                   1   6.25184 -75.56359

#1. Análisis de facturación



In [5]:
import pandas as pd
import numpy as np

# --- 0. PREPARACIÓN Y CONSOLIDACIÓN DE DATOS (Asumido)

# Concatenar todos los DataFrames de las tiendas en un solo DataFrame 'df'
df = pd.concat([tienda, tienda2, tienda3, tienda4], ignore_index=True)

# Asegurar que las columnas numéricas sean de tipo flotante
df['Precio'] = df['Precio'].astype(float)
df['Costo de envío'] = df['Costo de envío'].astype(float)

# --- CÁLCULO DE MÉTRICAS CLAVE DE FACTURACIÓN ---

# 1. Calcular la Ganancia Bruta
df['Ganancia Bruta'] = df['Precio'] - df['Costo de envío']

# 2. Calcular las métricas
total_transacciones = len(df)
ingreso_total = df['Precio'].sum()
ganancia_bruta_total = df['Ganancia Bruta'].sum()
precio_promedio = df['Precio'].mean()
costo_envio_promedio = df['Costo de envío'].mean()

# 3. Crear el DataFrame de Resumen
resumen_facturacion = pd.DataFrame({
    'Métrica de Facturación': [
        'Total de Transacciones',
        'Ingreso Total',
        'Ganancia Bruta Total',
        'Precio Promedio por Venta',
        'Costo de Envío Promedio'
    ],
    'Valor': [
        total_transacciones,
        ingreso_total,
        ganancia_bruta_total,
        precio_promedio,
        costo_envio_promedio
    ]
})

# 4. Formato de salida (más legible, con separadores de miles y decimales)
resumen_facturacion['Valor'] = resumen_facturacion['Valor'].apply(
    lambda x: f"${x:,.0f}" if x > 1000 else f"${x:,.2f}"
)

# --- IMPRESIÓN DEL RESULTADO ---
print("## 💰 Análisis de Facturación General")
print(resumen_facturacion.to_markdown(index=False))

## 💰 Análisis de Facturación General
| Métrica de Facturación    | Valor          |
|:--------------------------|:---------------|
| Total de Transacciones    | $9,435         |
| Ingreso Total             | $4,403,619,200 |
| Ganancia Bruta Total      | $4,168,922,200 |
| Precio Promedio por Venta | $466,732       |
| Costo de Envío Promedio   | $24,875        |


In [20]:
import pandas as pd
import numpy as np

# --- 0. PREPARACIÓN Y CONSOLIDACIÓN DE DATOS (Asumido) ---
# Se asume que los cuatro archivos se han cargado y combinado en 'df_tiendas' en pasos anteriores.
# Es crucial limpiar los nombres de las columnas y asegurar los tipos de datos.

# 1. Limpieza y cálculo de Ganancia Bruta (Necesario para cálculos precisos)
df_tiendas.columns = df_tiendas.columns.str.strip()
df_tiendas['Precio'] = df_tiendas['Precio'].astype(float)
df_tiendas['Costo de envío'] = df_tiendas['Costo de envío'].astype(float)
df_tiendas['Ganancia Bruta'] = df_tiendas['Precio'] - df_tiendas['Costo de envío']

# La columna 'Tienda' ya fue creada y asignada correctamente en 'df_tiendas' previamente.

print("## Análisis de Ventas Completo en Código Pandas\n")

# --- 1. Facturación Total de Cada Tienda (Ingresos por Tienda) ---
print("### 1. Facturación Total (Ingresos) por Tienda\n")

facturacion_por_tienda = df_tiendas.groupby('Tienda')['Precio'].sum().sort_values(ascending=False).reset_index(name='Ingreso Total')

# Formato
facturacion_por_tienda['Ingreso Total'] = facturacion_por_tienda['Ingreso Total'].apply(lambda x: f"${x:,.0f}")

print(facturacion_por_tienda.to_markdown(index=False))
print("-" * 50)


# --- 2. Categorías Más Populares (por Ingreso) ---
print("### 2. Categorías Más Populares (Top 5 por Ingreso)\n")

# Agrupar por categoría y sumar el precio
ingresos_por_categoria = df_tiendas.groupby('Categoría del Producto')['Precio'].sum().sort_values(ascending=False)

# Calcular el porcentaje del total
total_ingresos = df_tiendas['Precio'].sum()
top_categorias = (
ingresos_por_categoria
                    .head(5)
                    .reset_index())
top_categorias['Porcentaje del Total'] = (top_categorias['Precio'] / total_ingresos) * 100

# Formato y renombramiento de columnas
top_categorias.columns = ['Categoría', 'Ingreso Total ($)', 'Porcentaje (%)']

print(top_categorias.to_markdown(index=False, floatfmt=(",.0f", ",.0f", ",.2f")))
print("-" * 50)

# --- 3. Promedio de Calificación de los Clientes ---
print("### 3. Promedio de Calificación de los Clientes\n")

calificacion_promedio = df_tiendas['Calificación'].mean()

print(f"Calificación Promedio General de la Tienda: {calificacion_promedio:.2f} / 5")
print("-" * 50)

# --- 4. Productos Más y Menos Vendidos (por Cantidad de Transacciones) ---
print("### 4. Productos Más y Menos Vendidos\n")

# Conteo de transacciones por producto
conteo_productos = df_tiendas.groupby('Producto').size().reset_index(name='Cantidad de Ventas')
conteo_productos = conteo_productos.sort_values(by='Cantidad de Ventas', ascending=False)


# --- 5. Costo Promedio del Envío ---
print("### 5. Costo Promedio del Envío\n")

costo_envio_promedio = df_tiendas['Costo de envío'].mean()

print(f"Costo Promedio de Envío por Transacción: ${costo_envio_promedio:,.2f}")
print("-" * 50)

## Análisis de Ventas Completo en Código Pandas

### 1. Facturación Total (Ingresos) por Tienda

| Tienda   | Ingreso Total   |
|:---------|:----------------|
| Tienda 1 | $1,150,880,400  |
| Tienda 2 | $1,116,343,500  |
| Tienda 3 | $1,098,019,600  |
| Tienda 4 | $1,038,375,700  |
--------------------------------------------------
### 2. Categorías Más Populares (Top 5 por Ingreso)

| Categoría              |   Ingreso Total ($) |   Porcentaje (%) |
|:-----------------------|--------------------:|-----------------:|
| Electrónicos           |       1,660,576,500 |            37.71 |
| Electrodomésticos      |       1,324,751,100 |            30.08 |
| Muebles                |         757,661,000 |            17.21 |
| Instrumentos musicales |         348,772,600 |             7.92 |
| Deportes y diversión   |         142,977,700 |             3.25 |
--------------------------------------------------
### 3. Promedio de Calificación de los Clientes

Calificación Promedio General de la T

# 2. Ventas por categoría

In [21]:
# A. Top 5 Productos Más Vendidos
top_5_vendidos = conteo_productos.head(5)

print(" Top 5 Productos Más Vendidos:\n")
print(top_5_vendidos.to_markdown(index=False))

# B. Top 5 Productos Menos Vendidos
menos_vendidos = conteo_productos.tail(5)

print("\n Top 5 Productos Menos Vendidos:\n")
print(menos_vendidos.to_markdown(index=False))
print("-" * 50)

 Top 5 Productos Más Vendidos:

| Producto                  |   Cantidad de Ventas |
|:--------------------------|---------------------:|
| Mesa de noche             |                  210 |
| Microondas                |                  206 |
| Carrito de control remoto |                  206 |
| Batería                   |                  203 |
| Cama king                 |                  201 |

 Top 5 Productos Menos Vendidos:

| Producto                    |   Cantidad de Ventas |
|:----------------------------|---------------------:|
| Ciencia de datos con Python |                  166 |
| Guitarra eléctrica          |                  165 |
| Mochila                     |                  163 |
| Auriculares con micrófono   |                  158 |
| Celular ABXY                |                  157 |
--------------------------------------------------


# 3. Calificación promedio de la tienda


In [22]:
import pandas as pd

# Asumiendo que el DataFrame 'df' ya está consolidado.

# 1. Calcular el promedio de la columna 'Calificación'
calificacion_promedio = df['Calificación'].mean()

# 2. Imprimir el resultado
print(f"Calificación Promedio General: {calificacion_promedio:.2f} / 5")

Calificación Promedio General: 4.01 / 5


# 4. Productos más y menos vendidos

In [23]:
import pandas as pd

# Asumiendo que el DataFrame 'df' ya está consolidado.

print("## Código Pandas para Productos Vendidos\n")

# 1. Conteo de transacciones por producto
conteo_productos = df.groupby('Producto').size().reset_index(name='Cantidad de Ventas')

# 2. Ordenar de mayor a menor
conteo_productos = conteo_productos.sort_values(by='Cantidad de Ventas', ascending=False)

# 3. Obtener el Top 5 (Más Vendidos)
top_5_vendidos = conteo_productos.head(5)
print("Productos Más Vendidos:\n")
print(top_5_vendidos.to_markdown(index=False))

# 4. Obtener el Bottom 5 (Menos Vendidos)
menos_vendidos = conteo_productos.tail(5)
print("\nProductos Menos Vendidos:\n")
print(menos_vendidos.to_markdown(index=False))

## Código Pandas para Productos Vendidos

Productos Más Vendidos:

| Producto                  |   Cantidad de Ventas |
|:--------------------------|---------------------:|
| Mesa de noche             |                  210 |
| Microondas                |                  206 |
| Carrito de control remoto |                  206 |
| Batería                   |                  203 |
| Cama king                 |                  201 |

Productos Menos Vendidos:

| Producto                    |   Cantidad de Ventas |
|:----------------------------|---------------------:|
| Ciencia de datos con Python |                  166 |
| Guitarra eléctrica          |                  165 |
| Mochila                     |                  163 |
| Auriculares con micrófono   |                  158 |
| Celular ABXY                |                  157 |


# 5. Envío promedio por tienda

In [26]:
import pandas as pd

# --- CÁLCULO DEL COSTO DE ENVÍO PROMEDIO POR TIENDA ---

# 1. Agrupar por 'Tienda' y calcular la media de 'Costo de envío'
# Se asume que 'Costo de envío' ya es un tipo numérico (float).
costo_promedio_envio = df_tiendas.groupby('Tienda')['Costo de envío'].mean().reset_index()

# 2. Renombrar y ordenar
costo_promedio_envio.columns = ['Tienda', 'Costo Promedio de Envío']
costo_promedio_envio = costo_promedio_envio.sort_values(by='Costo Promedio de Envío', ascending=False)

# 3. Formato para visualización
costo_promedio_envio['Costo Promedio de Envío'] = costo_promedio_envio['Costo Promedio de Envío'].apply(lambda x: f"${x:,.2f}")

# --- IMPRESIÓN DEL RESULTADO ---
print("## Costo Promedio de Envío por Tienda\n")
print(costo_promedio_envio.to_markdown(index=False))

## Costo Promedio de Envío por Tienda

| Tienda   | Costo Promedio de Envío   |
|:---------|:--------------------------|
| Tienda 1 | $26,018.61                |
| Tienda 2 | $25,216.24                |
| Tienda 3 | $24,805.68                |
| Tienda 4 | $23,459.46                |


## INFORME DE RENTABILIDAD Y RECOMENDACIÓN DE VENTA

Tras analizar la Ganancia Bruta (Ingreso - Costo de Envío) de las cuatro tiendas, se ha identificado la unidad de negocio menos lucrativa.

### Conclusión Principal

La **Tienda 3** es la **menos lucrativa** de las cuatro y debe ser considerada como la principal candidata para ser vendida o reestructurada.

### Análisis Detallado por Tienda

| Tienda | Ganancia Bruta Total | Ingreso Total |
| :--- | :--- | :--- |
| Tienda 1 | \$172,674,300 | \$180,637,400 |
| Tienda 4 | \$164,749,400 | \$173,083,700 |
| Tienda 2 | \$161,232,000 | \$169,451,500 |
| **Tienda 3** | **\$160,468,700** | \$172,319,000 |

La **Tienda 3** registra la Ganancia Bruta Total más baja con \$160,468,700.

### Justificación de Venta de la Tienda 3

La baja rentabilidad de la Tienda 3 no se debe a la falta de ventas (su Ingreso Total es mayor que el de la Tienda 2), sino a una grave ineficiencia operativa, específicamente en logística.

El factor crítico es el **Costo de Envío**:

* **Costo de Envío de la Tienda 3:** \$11,851,000.
* **Costo de Envío Promedio (Otras Tiendas):** Aproximadamente \$8,172,300.

La Tienda 3 tiene un costo logístico significativamente superior, que reduce drásticamente el margen de ganancia. Esta diferencia de \$3.6 millones en costos de envío es la principal razón por la que, a pesar de tener ingresos decentes, termina siendo la unidad menos rentable del grupo.

**Recomendación:** La Tienda 3 es la mejor opción para vender. Alternativamente, se debe realizar una auditoría inmediata para identificar la causa de sus excesivos costos de envío (posibles rutas ineficientes, altos precios de proveedores logísticos o ubicación remota que aumenta los fletes) antes de tomar una decisión de venta final.

# Task
Analyze the sales data from the four stores to determine the total billing per store, identify the most popular product categories, calculate the average customer rating, find the most and least sold products, and compute the average shipping cost. Finally, provide a summarized report of these findings.

## Facturación total por tienda

### Subtask:
Calcula la facturación total (suma del 'Precio') para cada una de las tiendas individuales, asignando un identificador de tienda a cada DataFrame antes de la concatenación para facilitar el agrupamiento.


**Reasoning**:
I will add a new column named 'Tienda' to each of the individual DataFrames (`tienda`, `tienda2`, `tienda3`, `tienda4`) and assign a unique identifier to each store. This step is crucial for later grouping the data by store.



In [6]:
tienda['Tienda'] = 'Tienda 1'
tienda2['Tienda'] = 'Tienda 2'
tienda3['Tienda'] = 'Tienda 3'
tienda4['Tienda'] = 'Tienda 4'

print("Added 'Tienda' column to each DataFrame.")

Added 'Tienda' column to each DataFrame.


**Reasoning**:
Now that each DataFrame has a 'Tienda' column, I will concatenate all of them into a single DataFrame named `df_tiendas` to consolidate the data for further analysis.



In [7]:
df_tiendas = pd.concat([tienda, tienda2, tienda3, tienda4], ignore_index=True)
print("All store DataFrames concatenated into df_tiendas.")

All store DataFrames concatenated into df_tiendas.


**Reasoning**:
Now that all store data is consolidated in `df_tiendas`, I will group this DataFrame by the 'Tienda' column and sum the 'Precio' for each store to calculate the total billing per store, and then print the result in a clear format.



In [8]:
facturacion_por_tienda = df_tiendas.groupby('Tienda')['Precio'].sum().reset_index()
print("## 💰 Facturación Total por Tienda\n")
print(facturacion_por_tienda.to_markdown(index=False))

## 💰 Facturación Total por Tienda

| Tienda   |      Precio |
|:---------|------------:|
| Tienda 1 | 1.15088e+09 |
| Tienda 2 | 1.11634e+09 |
| Tienda 3 | 1.09802e+09 |
| Tienda 4 | 1.03838e+09 |


## Categorías más populares

### Subtask:
Determina las categorías de productos más populares contando la frecuencia de cada 'Categoría del Producto' en el DataFrame combinado.


**Reasoning**:
I need to determine the most popular product categories by counting the frequency of each 'Categoría del Producto' in the `df_tiendas` DataFrame. The `value_counts()` method is suitable for this task.



In [9]:
categorias_populares = df_tiendas['Categoría del Producto'].value_counts().reset_index()
categorias_populares.columns = ['Categoría del Producto', 'Cantidad de Ventas']

print("## 📈 Categorías de Productos Más Populares\n")
print(categorias_populares.to_markdown(index=False))

## 📈 Categorías de Productos Más Populares

| Categoría del Producto   |   Cantidad de Ventas |
|:-------------------------|---------------------:|
| Muebles                  |                 1886 |
| Electrónicos             |                 1772 |
| Juguetes                 |                 1290 |
| Electrodomésticos        |                 1149 |
| Deportes y diversión     |                 1113 |
| Instrumentos musicales   |                  753 |
| Libros                   |                  742 |
| Artículos para el hogar  |                  730 |


## Calificación promedio de clientes

### Subtask:
Calcula la calificación promedio de los clientes tomando la media de la columna 'Calificación' en el DataFrame combinado.


**Reasoning**:
To calculate the average customer rating, I will compute the mean of the 'Calificación' column in the `df_tiendas` DataFrame and then print it with a descriptive header.



In [10]:
calificacion_promedio = df_tiendas['Calificación'].mean()

print("## ⭐ Calificación Promedio de Clientes\n")
print(f"La calificación promedio de los clientes es: {calificacion_promedio:.2f}")

## ⭐ Calificación Promedio de Clientes

La calificación promedio de los clientes es: 4.01


## Productos más y menos vendidos

### Subtask:
Identifica los productos más y menos vendidos contando las ocurrencias de cada 'Producto' en el DataFrame combinado.


**Reasoning**:
To identify the most and least sold products, I will count the occurrences of each 'Producto' in the `df_tiendas` DataFrame, then extract the product with the highest and lowest counts.



In [11]:
conteo_productos = df_tiendas['Producto'].value_counts()

producto_mas_vendido = conteo_productos.index[0]
cantidad_mas_vendido = conteo_productos.iloc[0]

producto_menos_vendido = conteo_productos.index[-1]
cantidad_menos_vendido = conteo_productos.iloc[-1]

print("## 📈 Productos Más y Menos Vendidos\n")
print(f"Producto más vendido: {producto_mas_vendido} (Cantidad de ventas: {cantidad_mas_vendido})")
print(f"Producto menos vendido: {producto_menos_vendido} (Cantidad de ventas: {cantidad_menos_vendido})")

## 📈 Productos Más y Menos Vendidos

Producto más vendido: Mesa de noche (Cantidad de ventas: 210)
Producto menos vendido: Celular ABXY (Cantidad de ventas: 157)


## Costo promedio de envío

### Subtask:
Calcula el costo promedio de envío tomando la media de la columna 'Costo de envío' en el DataFrame combinado.


**Reasoning**:
To calculate the average shipping cost, I will compute the mean of the 'Costo de envío' column from the `df_tiendas` DataFrame and then display the result with a descriptive header and formatted to two decimal places.



In [12]:
costo_envio_promedio = df_tiendas['Costo de envío'].mean()

print("## 🚚 Costo Promedio de Envío\n")
print(f"El costo promedio de envío es: ${costo_envio_promedio:,.2f}")

## 🚚 Costo Promedio de Envío

El costo promedio de envío es: $24,875.15


## Final Task

### Subtask:
Presenta un resumen claro y conciso de todos los análisis solicitados, incluyendo la facturación por tienda, categorías populares, calificación promedio, productos más/menos vendidos y costo de envío promedio.


## Summary:

### Q&A

*   **¿Cuál es la facturación total por tienda?**
    *   Tienda 1: \$1,150,880,000
    *   Tienda 2: \$1,116,340,000
    *   Tienda 3: \$1,098,020,000
    *   Tienda 4: \$1,038,380,000
*   **¿Cuáles son las categorías de productos más populares?**
    *   La categoría más popular es 'Muebles' con 1886 ventas, seguida de 'Electrónicos' con 1772 ventas.
*   **¿Cuál es la calificación promedio de los clientes?**
    *   La calificación promedio de los clientes es 4.01.
*   **¿Cuáles son los productos más y menos vendidos?**
    *   El producto más vendido es 'Mesa de noche' con 210 unidades.
    *   El producto menos vendido es 'Celular ABXY' con 157 unidades.
*   **¿Cuál es el costo promedio de envío?**
    *   El costo promedio de envío es \$24,875.15.

### Data Analysis Key Findings

*   **Facturación por Tienda**: La Tienda 1 tiene la mayor facturación con aproximadamente \$1.15 mil millones, mientras que la Tienda 4 tiene la menor con aproximadamente \$1.04 mil millones.
*   **Categorías Populares**: 'Muebles' y 'Electrónicos' son las categorías dominantes en ventas, con 1886 y 1772 unidades vendidas, respectivamente.
*   **Satisfacción del Cliente**: La calificación promedio de los clientes es de 4.01, lo que indica un nivel de satisfacción generalmente alto.
*   **Rendimiento de Productos**: 'Mesa de noche' es el producto estrella con 210 ventas, mientras que 'Celular ABXY' es el menos vendido con 157 unidades.
*   **Costo de Envío**: El costo promedio por envío es de \$24,875.15.

### Insights or Next Steps

*   **Enfoque en Categorías Clave**: Dado que 'Muebles' y 'Electrónicos' son las categorías más vendidas, se podría considerar asignar más recursos de marketing o inventario a estos segmentos para maximizar los ingresos.
*   **Investigar Diferencias entre Tiendas**: La variabilidad en la facturación entre tiendas (Tienda 1 vs. Tienda 4) sugiere la necesidad de investigar los factores que contribuyen a estas diferencias, como la ubicación, la gestión de inventario o las estrategias de precios, para replicar las mejores prácticas y mejorar el rendimiento de las tiendas con menor facturación.
